zerver/lib/streams.py

```diff
@@ -1102,6 +1140,36 @@ def can_administer_accessible_channel(channel: Stream, user_profile: UserProfile
     )
 
 
+def get_metadata_access_streams(
+    user_profile: UserProfile,
+    streams: Collection[Stream],
+    user_group_membership_details: UserGroupMembershipDetails,
+) -> list[Stream]:
+    if len(streams) == 0:
+        return []
+
+    recipient_ids = [stream.recipient_id for stream in streams]
+    subscribed_recipient_ids = set(
+        Subscription.objects.filter(
+            user_profile=user_profile, recipient_id__in=recipient_ids, active=True
+        ).values_list("recipient_id", flat=True)
+    )
+
+    metadata_access_streams: list[Stream] = []
+
+    for stream in streams:
+        is_subscribed = stream.recipient_id in subscribed_recipient_ids
+        if user_has_metadata_access(
+            user_profile,
+            stream,
+            user_group_membership_details,
+            is_subscribed=is_subscribed,
+        ):
+            metadata_access_streams.append(stream)
+
+    return metadata_access_streams
```

Comment:
This implementation that fetches all the channels and then filters is probably not ideal for some situations where an organization has thousands of channels and the user can only access like 5 of them. But a faster implementation requires doing the permissions logic more in SQL, and this isn't likely to be used often in that scenario, so I think we should just go with this version.